### PR TITLE
Random cleanup in NuGet restore code

### PR DESCRIPTION
**Pull request by davkean**
Pulled out self-contained changes that I've made in the rewrite of the NuGet restore code to make review purposes easier.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/RestoreBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+using NuGet.SolutionRestoreManager;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.NuGet
+{
+    /// <summary>
+    ///     Contains builder methods for creating <see cref="IVsProjectProperty"/> and 
+    ///     <see cref="IVsReferenceItems"/> instances.
+    /// </summary>
+    internal class RestoreBuilder
```
Comment (davidwengier) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/RestoreBuilder.cs]: nit: static?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ProjectReference.xaml [hunk 2]
```diff
@@ -66,9 +66,17 @@
                   DisplayName="Exclude Assets"
                   Visible="True" />
 
-  <StringProperty Name="FullPath"
-                  Visible="False" />
-  
+  <StringProperty Name="ProjectFileFullPath"
```
Comment (davidwengier) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ProjectReference.xaml]: Nit: We should keep these files in alphabetical order for easier comparisons

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ProjectReference.xaml]: 👍

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/RestoreBuilder.cs (hunk 1 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/RestoreBuilder.cs]: 👍

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs [hunk 3]
```diff
@@ -82,30 +82,30 @@ internal static IVsProjectRestoreInfo Build(IEnumerable<IProjectVersionedValue<I
...
+                    toolReferences.Values
                 )
                 : null;
         }
 
-        private static IVsProjectProperties GetProperties(IImmutableDictionary<string, string> items)
+        private static IEnumerable<IVsProjectProperty> GetProperties(IImmutableDictionary<string, string> items)
         {
-            return new ProjectProperties(items.Select(v => new ProjectProperty(v.Key, v.Value)));
+            return items.Select(v => new ProjectProperty(v.Key, v.Value));
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs]: I'm not sure why this is better?  It makes it unclear whether this is a deferred enumerator (as this is), and if that's okay.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ProjectReference.xaml (hunk 2 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/ProjectReference.xaml]: @davidwengier this is another candidate for the unit tests.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/ReferenceItem.cs [hunk 4]
```diff
@@ -1,18 +1,23 @@
...
     {
-        public ReferenceItem(string name, IVsReferenceProperties properties)
+        public ReferenceItem(string name, IEnumerable<IVsReferenceProperty> properties)
         {
             Requires.NotNullOrEmpty(name, nameof(name));
             Requires.NotNull(properties, nameof(properties));
 
             Name = name;
-            Properties = properties;
+            Properties = new ReferenceProperties(properties);
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/ReferenceItem.cs]: Why introduce this logic here? It means the type is no longer a simple 'record'. Can `RestoreBuilder`'s methods just return `IVsReferenceProperties` and `IVsReferenceItems`? Does this make testing in your subsequent PR easier?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/TargetFrameworkInfo.cs [hunk 5]
```diff
@@ -1,22 +1,27 @@
...
             Requires.NotNull(packageReferences, nameof(packageReferences));
             Requires.NotNull(properties, nameof(properties));
 
             TargetFrameworkMoniker = targetFrameworkMoniker;
-            ProjectReferences = projectReferences;
-            PackageReferences = packageReferences;
-            Properties = properties;
+            ProjectReferences = new ReferenceItems(projectReferences);
+            PackageReferences = new ReferenceItems(packageReferences);
+            Properties = new ProjectProperties(properties);
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/TargetFrameworkInfo.cs]: Ditto.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/ReferenceItem.cs (hunk 4 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/ReferenceItem.cs]: It simplifies construction/merging in a later PR. It maps very strongly to other parallel concepts; you take an IEnumerable<T> and map to strongly-typed collections (such as Collection<T>/ReadOnlyCollection<T>).

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs (hunk 3 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs]: You and @drewnoakes had similar feedback, I've reverted back to using the strongly typed collections.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/ReferenceItem.cs (hunk 4 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/Snapshots/ReferenceItem.cs]: I've reverted this back, both you @Pilchie had similar feedback.

**Pull request merged**